File: DSA_2215/stack/prac/main.c

#include <stdio.h>

int menu();
void push(int *, int *, int);
void pop(int *, int *);
void display(int *, int *);
void pushMutiple(int *stack, int *top, int size);
void justPush(int *stack, int *top, int size, int data);

int main(int argc, char const *argv[]) {
  int size;
  int stack[100];

  for (;;) {
    switch (menu()) {
      case 1:
        push(stack, &size, 100);
        break;
      case 2:
        pop(stack, &size);
        break;
      case 3:
        display(stack, &size);
        break;
      case 4:
        return 0;
      case 5:
        pushMutiple(stack, &size, 100);
      default:
        printf("Invalid choice\n");
    }
  }

  return 0;
}

int menu() {
  int choice;

  printf("1. Push\n");
  printf("2. Pop\n");
  printf("3. Display\n");
  printf("4. Exit\n");
  printf("5. Push multiple\n");
  printf("Enter your choice: ");
  scanf("%d", &choice);
  return choice;
}

void pop(int *stack, int *top) {
  if (*top == -1) {
    printf("Stack is empty\n");
    return;
  }
  printf("Popped element: %d\n", stack[(*top)--]);
}

void display(int *stack, int *top) {
  if (*top == -1) {
    printf("Stack is empty\n");
    return;
  }
  printf("Stack: ");
  for (int i = 0; i <= *top; i++) {
    printf("%d ", stack[i]);
  }
  printf("\n");
}

void push(int *stack, int *top, int size) {
  if (*top == size - 1) {
    printf("Stack is full\n");
    return;
  }
  int data;
  printf("Enter data: ");
  scanf("%d", &data);
  stack[++(*top)] = data;
}

void justPush(int *stack, int *top, int size, int data) {
  if (*top == size - 1) {
    printf("Stack is full\n");
    return;
  }
  stack[++(*top)] = data;
}

void pushMutiple(int *stack, int *top, int size) {
  if (*top == size - 1) {
    printf("Stack is full\n");
    return;
  }

  int n;
  printf("Enter number of elements: ");
  scanf("%d", &n);

  if (*top + n >= size) {
    printf("Stack is full\n");
    return;
  }

  printf("Enter data: ");
  for (int i = 0; i < n; i++) {
    // scanf("%d", &stack[++(*top)]);
    int data;
    scanf("%d", &data);
    justPush(stack, top, size, data);
  }
}

File: DSA_2215/stack/prac/main.go
package main

import "fmt"

func main() {
	stack := newStack()
	for {
		switch menu() {
		case 1:
			var value int
			fmt.Print("Enter value: ")
			fmt.Scanln(&value)
			stack.push(value)
		case 2:
			fmt.Println(stack.pop())
		case 3:
			fmt.Println(stack.size())
		case 4:
			fmt.Println(stack.top())
		case 5:
			fmt.Println(stack.data)
		case 6:
			return

		default:
			fmt.Println("Invalid choice")
		}
	}
}

func menu() int {
	fmt.Println("1. Push")
	fmt.Println("2. Pop")
	// stack size
	fmt.Println("3. Stack size")
	fmt.Println("4. top element")
	fmt.Println("5. Display stack")
	fmt.Println("6. Exit")
	fmt.Print("Enter your choice: ")

	var choice int
	fmt.Scanln(&choice)
	return choice
}

type stack struct {
	data []int
}

func newStack() *stack {
	return &stack{}
}

func (s *stack) push(value int) {
	s.data = append(s.data, value)
}

func (s *stack) pop() int {
	if len(s.data) == 0 {
		return -1
	}
	value := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return value
}

func (s *stack) size() int {
	return len(s.data)
}

func (s *stack) top() int {
	if len(s.data) == 0 {
		return -1
	}
	return s.data[len(s.data)-1]
}
